function: capture metric appearance for xrate/xincrease/xdelta

GiedriusS opened the pull request:
Current check for zero value injection doesn't make sense - we need to check it for since the beginning of when the metric appears.

GiedriusS commented:
@mklocke-shopify what do you think about this as you've worked on the original implementation in this engine?

fpetkovski commented on execution/scan/matrix_selector.go:
```diff
@@ -32,12 +32,13 @@ type matrixScanner struct {
+	storage        engstore.SeriesSelector
+	call           function.FunctionCall
+	scanners       []matrixScanner
+	metricAppeared []int64
```
Should we put this in the scanner itself? Also does it denote the time when the metric first appeared?

fpetkovski commented on execution/scan/matrix_selector.go:
```diff
@@ -155,13 +156,14 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
-				Offset:      o.offset,
-			})
+			fa := function.NewFunctionsArgs()
+			fa.Labels = series.labels
```
Does it make sense to have a constructor with all parameters so we don't accidentally omit something?

mklocke-shopify reviewed (approved):
This looks good! Thanks a lot for adding this, it makes sense.

GiedriusS commented:
Thx for the review @fpetkovski, I have refactored the code to be more Go-like:

- Instead of `-1` to indicate the absence of a value, let's use a pointer to int64; this makes the zero value of FunctionArgs and others make sense;
- Moved `metricAppeared` into scanners;
- Appended `ts` suffix everywhere so that it would be clear that it stores a timestamp.

fpetkovski commented on execution/scan/matrix_selector.go:
```diff
@@ -366,10 +368,20 @@ loop:
 		case chunkenc.ValNone:
 			break loop
 		case chunkenc.ValHistogram, chunkenc.ValFloatHistogram:
+			t, h := buf.AtHistogram()
```
Hmm, we call `AtFloatHistogram` bellow, so I wonder why we need to call `AtHistogram` here. We also seem to now append to the output before we check for staleness, so we might end up picking up stale samples. Is this the correct behavior?

GiedriusS replied:
You are correct, this was an artifact of rebasing. I somehow missed this part. Thank you for a diligent review :bow:
